sync/optimistic.md

```diff
@@ -0,0 +1,340 @@
+these conditions.*
+
+### How to optimistically import blocks
+
+To optimistically import a block:
+
+- The [`execute_payload`](../specs/bellatrix/beacon-chain.md#execute_payload) function MUST return `True` if the execution
+	engine returns `SYNCING` or `VALID`. An `INVALID` response MUST return `False`.
+- The [`validate_merge_block`][] function MUST NOT raise an assertion if both the
+`pow_block` and `pow_parent` are unknown to the execution engine.
+- The parent of the block MUST NOT have an INVALID execution payload.
+
+In addition to this change to validation, the consensus engine MUST be able to
+ascertain, after import, which blocks returned `SYNCING` and which returned
+`VALID`.
+
+Optimistically imported blocks MUST pass all verifications included in
+`process_block` (withstanding the modifications to `execute_payload`).
+
+A consensus engine MUST be able to retrospectively (i.e., after import) modify
+the status of `SYNCING` blocks to be either `VALID` or `INVALID` based upon responses
+from an execution engine. I.e., perform the following transitions:
+
+- `SYNCING` -> `VALID`
+- `SYNCING` -> `INVALID`
+
+When a block transitions from `SYNCING` -> `VALID`, all *ancestors* of the
+block MUST also transition from `SYNCING` -> `VALID`. Such a block is no longer
+considered "optimistically imported".
+
+When a block transitions from `SYNCING` -> `INVALID`, all *descendants* of the
+block MUST also transition from `SYNCING` -> `INVALID`.
+
+When a block transitions from the `SYNCING` state it is removed from the set of
+`store.optimistic_roots`.
+
+When a "merge block" (i.e. a block which enables execution) is declared to be
+`VALID` by an execution engine (either directly or indirectly), the full
+[`validate_merge_block`][] MUST be run against the merge block. If the block
+fails [`validate_merge_block`][], the merge block MUST be treated the same as
+an `INVALID` block (i.e., it and all its descendants are invalidated and
+removed from the block tree).
+
+### Execution Engine Errors
+
+When an execution engine returns an error or fails to respond to a payload
+validity request for some block, a consensus engine:
+
+- MUST NOT optimistically import the block.
+- MUST NOT apply the block to the fork choice store.
+- MAY queue the block for later processing.
+
+### Assumptions about Execution Engine Behaviour
+
+This specification assumes execution engines will only return `SYNCING` when
+there is insufficient information available to make a `VALID` or `INVALID`
+determination on the given `ExecutionPayload` (e.g., the parent payload is
+unknown). Specifically, `SYNCING` responses should be fork-specific, in that
+the search for a block on one chain MUST NOT trigger a `SYNCING` response for
+another chain.
+
+### Re-Orgs
+
+The consensus engine MUST support any chain reorganisation which does *not*
+affect the justified checkpoint. The consensus engine MAY support re-orgs
```

Comment:
Can this section clarify these are re-orgs of optimistic chain segments?

A fully verified justification will never revert if the consensus engine is following the spec *and* there is no 2/3 slashing event. (but a 2/3 slashing even could lead to a finality that is discordant with your current justification and pull you to another chain)

Reply:
Oh yeah, good point. In f4a125c8 I just removed this second sentence since I don't think it was really giving us anything. LMK what you think.

Reply:
works for me